modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/TextFieldBehavior.java

```diff
@@ -183,9 +183,10 @@ public void setTextFieldSkin(TextFieldSkin skin) {
 
         textField.commitValue();
         textField.fireEvent(actionEvent);
-
-        if (onAction == null && !actionEvent.isConsumed()) {
-            forwardToParent(event);
+        // fix of JDK-8207759: reverted logic
+        // mapping not auto-consume and consume if handled by action
+        if (onAction != null || actionEvent.isConsumed()) {
+            event.consume();
         }
```

Comment:
@kleopatra @kevinrushforth @aghaisas 

Shouldn't this be an `&&` ?  Now having an empty `setOnAction` that doesn't consume anything (but just logs for example) will affect the operation of this control.

Reply:
correct, this is a bug.
#1523

Reply:
hmm .. it's been a while but you might be on to something. 

Let's recap:

- the first issue to get fixed was JDK-8207774: the problem was that the isConsumed always was false (due to creating the actionEvent for passing around was done by the constructor that led to immediate copying, setting consumed to false)
- the next issue was this: the problem was that actively refiring the keyEvent on parent led to a nested eventDispatch, confusing every collaborator. That was fixed by removing the refire altogether and consuming the keyEvent if handled by the textField's action/Listeners (that's the plainly reverted logic above)
- both before and after the fixes the state "handled" can be reached by the mere existence of an onAction handler on the textField, consuming or not

And that's still wrong, as you correctly - IMO - noted. The expectation formulated as a test (c&p and adjusted from TextFieldTest)

    /**
     * ENTER must be forwarded if action
     * not consumed the action.
     *
     * Here we test that key handlers on parent are notified.
     */
    @Test
    public void testEnterWithNotConsumingActionParentHandler() {
        initStage();
        root.getChildren().add(txtField);
        txtField.setOnAction(e -> {}); // do nothing handler
        List<KeyEvent> keys = new ArrayList<>();

        root.addEventHandler(KeyEvent.KEY_PRESSED, e -> {
            keys.add(e);
        });
        stage.show();
        KeyEventFirer keyboard = new KeyEventFirer(txtField);
        keyboard.doKeyPress(ENTER);
        assertEquals("key handler must be notified", 1, keys.size());
    }

We can't reach that by replacing the || by && (would introduce a regression, as failing tests indicate) but .. maybe by entirely removing the not/null check of onAction: the fired actionEvent will be passed to the singleton handler (along with any added via addXX) anyway, so it doesn't really make sense (with the first issue fixed)

Now the method looks like and all tests including the new one above (and hopefully the analogous twins of XXConsuming -> XXNotConsuming, didn't try, being lazy ;):

    @Override protected void fire(KeyEvent event) {
        TextField textField = getNode();
        EventHandler<ActionEvent> onAction = textField.getOnAction();
        // use textField as target to prevent immediate copy in dispatch
        ActionEvent actionEvent = new ActionEvent(textField, textField);

        textField.commitValue();
        textField.fireEvent(actionEvent);
        if (actionEvent.isConsumed()) {
            event.consume();
        }
    }

What do you think? My code brain is a bit rusty, so might have missed something.

Reply:
@kleopatra Your explanation makes sense to me. Thank you for taking the time to write it up.

If so, the proposed fix, including adjusting the comment, would be:

```
--- a/modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/TextFieldBehavior.java
+++ b/modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/TextFieldBehavior.java
@@ -162,9 +162,8 @@ public class TextFieldBehavior extends TextInputControlBehavior<TextField> {

         textField.commitValue();
         textField.fireEvent(actionEvent);
-        // fix of JDK-8207759: reverted logic
-        // mapping not auto-consume and consume if handled by action
-        if (onAction != null || actionEvent.isConsumed()) {
+        // Consume the original event if the copied actionEvent was consumed.
+        if (actionEvent.isConsumed()) {
             event.consume();
         }
     }
```

Reply:
You are right, `&&` would be insufficient as there is another way to add action handlers.  I didn't really see that, althoug the fix I mind was to eliminate the `onAction` variable completely (it is unused in your fix as well), so probably would have ended up at the correct result :)

I can roll this into a small PR if you wish.

Reply:
FWIW, checking the isConsumed flag will only work if the action is consumed by a handler registered directly with the TextField. If there's a filter or handler attached to some other part of the scene graph a copy of the event will be made and consumed and `actionEvent.isConsumed()` will return false. A more complete fix is to replace `textField.fireEvent` with a call to `EventUtil.fireEvent` and check for a `null` return value.

If I understand this thread correctly if a KeyEvent generates an ActionEvent the KeyEvent should only be consumed if the ActionEvent is. If so the behavior should be consistent across TextFields and the controls that manage TextFields internally (Spinner, ComboBox, and DatePicker). I'm already working on a test for Escape key handling for this set of controls (for Escape the issue is whether a formatter is attached or not). I'm willing to write up a similar test for Enter handling for the same set of controls. I just need to make sure I understand exactly what behavior is expected.

Reply:
> FWIW, checking the isConsumed flag will only work if the action is consumed by a handler registered directly with the TextField. If there's a filter or handler attached to some other part of the scene graph a copy of the event will be made and consumed and `actionEvent.isConsumed()` will return false. A more complete fix is to replace `textField.fireEvent` with a call to `EventUtil.fireEvent` and check for a `null` return value.

It seems that an earlier check I did to see if events are copied wasn't done correctly.  Indeed events are copied at every filter/handler.  That does seem excessive.  It also makes the public `Node#fireEvent` a bit useless for proper event propagation for normal users that don't have access to `EventUtil`.

I checked the events that are copied, and it is basically making the copy to adjust the source (target remains constant).  That's really odd; I'd expect the source to remain the same as well (ie. the Scene or Window, not the "previous" filter/handler Node level), especially as it is documented as:

    /**
     * The object on which the Event initially occurred.
     *
     * @return the object on which the Event initially occurred
     */
    public Object getSource() {
        return source;
    }

Reply:
> I checked the events that are copied, and it is basically making the copy to adjust the source (target remains constant). That's really odd; I'd expect the source to remain the same as well (ie. the Scene or Window, not the "previous" filter/handler Node level), especially as it is documented as:

I haven't walked through this in AWT (there are only so many hours in the day) but I believe the AWT `source` corresponds to the JavaFX `target` i.e. the object the event was fired against. JavaFX repurposes `source` as the object the handler or filter was registered on. JavaFX shouldn't be pointing at the AWT documentation since it doesn't match the JavaFX implementation.

Reply:
> > I checked the events that are copied, and it is basically making the copy to adjust the source (target remains constant). That's really odd; I'd expect the source to remain the same as well (ie. the Scene or Window, not the "previous" filter/handler Node level), especially as it is documented as:
> 
> I haven't walked through this in AWT (there are only so many hours in the day) but I believe the AWT `source` corresponds to the JavaFX `target` i.e. the object the event was fired against. JavaFX repurposes `source` as the object the handler or filter was registered on. JavaFX shouldn't be pointing at the AWT documentation since it doesn't match the JavaFX implementation.

Well, it implements AWT's `EventObject` through `class Event extends EventObject` -- it would be rather silly to repurpose `getSource` for something else, but then still extend AWT's `EventObject`. 

The class docs of `Event` mention an event source, and describes it as "The event source specifies for an event handler the object on which that handler has been registered and which sent  the event to it" -- that sentence is rather hard to decode, but I guess their meaning is that the Event received by your event handler will have its source set to the object where you registered your handler (ie. a `Node` in most cases) -- it would be similar to `getBean` for properties, but in an far more roundabout way.

So basiscally, there are now two definitions of source, one that is constant, and one that requires an event to be copied before having an event filter or handler receive it...